include/LinearSystem.h

```diff
@@ -66,6 +66,16 @@ class CoeffApplier
 
   virtual void free_device_pointer() = 0;
   virtual CoeffApplier* device_pointer() = 0;
+  virtual void resetInternalData() {};
+  virtual void applyDirichletBCs(Realm &,
+				 stk::mesh::FieldBase *,
+				 stk::mesh::FieldBase *,
+				 const stk::mesh::PartVector&) {};
+  virtual void finishAssembly(void *, std::vector<void *>) {};
```

Comment:
I prefer not to see function signatures with un-named arguments. What are these arguments? And if I did know what they were, I would probably object to using void.

Reply:
unnamed gets rid of compiler warnings for derived classes that don't implement these methods like Tpetra. void * is nasty, I would agree. The challenge here is that I need to get my Kokkos data structures for the CRS matrix into a Hypre data structure. Now, I can't pass a Hypre specific type into this method as would break a non Hypre build. The alternative to passing Hypre structures in is to pull Kokkos CRS data structures out.

Reply:
ok, I see the challenge, I agree that's hard. Maybe you need to stick with void. As for the names, maybe you could put commented-out names, like finishAssembly(void * /\*my-arg\*/)

Reply:
yes. That's good. I'll try to stick to that convention.

Reply:
@PaulMullowney I don't see any reason why `FinishAssembly` needs to be part of the public API and therefore `LinearSystem`. In that case `finish_assembly` can just be a local method with its proper type for each LinearSystem that uses it.

Reply:
I may have tried to be local to HypreLinSysCoeffApplier and failed to get it to compile. I can try again.

Reply:
Actually this pertains to all the new methods of LinearSystem: resetInternalData, finishAssembly, sum_into_nonNGP, and applyDirichletBCs.

Reply:
I've tried:

HypreLinSysCoeffApplier* hcApplier = reinterpret_cast<HypreLinSysCoeffApplier*>(hostCoeffApplier);
hcApplier->finishAssembly(...);

This gives:../src/HypreLinearSystem.C:1194:99: error: invalid cast from type ‘std::unique_ptr<sierra::nalu::CoeffApplier>’ to type ‘sierra::nalu::HypreLinearSystem::HypreLinSysCoeffApplier*’
   HypreLinSysCoeffApplier* hcApplier = reinterpret_cast<HypreLinSysCoeffApplier*>(hostCoeffApplier);

What's the right way to do this?

Reply:
You need to extract the raw pointer from the unique_ptr, then use dynamic_cast. Very rarely use reinterpret_cast for anything. Maybe something like dynamic_cast<HypreLinSysCoeffApplier*>(hostCoeffApplier.get())

Reply:
You can also do 

```C++
auto coeff_applier = dynamic_cast<HypreLinSysCoeffApplier&>(*hostCoeffApplier);
```

Reply:
Ok. I think I've fixed it. I removed the methods from LinearSystem and used dynamic cast to call them in HypreLinearSystem. I'm gonna push shortly.

Reply:
methods are gone from the base class